[Upgrade] PostgreSQL 13 (#2468)
epiphany-platform/epiphany: docs/home/howto/UPGRADE.md

it currently can upgrade and will add are:

---
**NOTE**

Starting from v0.7.1 to the latest, Epiphany provides an ability to install K8s v1.18.6 only. There is an assertion to
check whether K8s version is supported before running upgrade.

---

- Kubernetes (master and nodes). At the moment only v1.18.6 is supported
- common: Upgrades all common configurations to match them to current Epiphany version
- repository: Adds the repository role needed for component installation in current Epiphany version
- image_registry: Adds the image_registry role needed for offline installation in current Epiphany version

The component upgrade takes the existing Ansible build output and based on that performs the upgrade of the currently
supported components. If you need to re-apply your entire Epiphany cluster a **manual** adjustment of the input yaml is
needed to the latest specification which then should be applied with `epicli apply...`. Please
see [Run apply after upgrade](./UPGRADE.md#run-apply-after-upgrade) chapter for more details.

Note about upgrade from pre-0.8 Epiphany:

> - If you run upgrade from version older than 0.8 you should make sure that you've got enough disk space on master (which

Reviewer: ```suggestion
- If you need to upgrade a cluster deployed with `epicli` in version earlier than 0.8, you should make sure that you've got enough disk space on master (which
```

  is used as repository) host. If you didn't extend OS disk on master during deployment process, you probably have only

>   32 GB disk which is not enough to properly upgrade cluster (we recommend 50 GB). Before you run upgrade please extend

Reviewer: ```suggestion
  32 GB disk which is not enough to properly upgrade cluster (we recommend at least 64 GB). Before you run upgrade, please extend
```

>   os disks on master machine according to cloud provider

Reviewer: ```suggestion
  OS disk on master machine according to cloud provider
```

  documentation: [AWS](https://aws.amazon.com/premiumsupport/knowledge-center/expand-root-ebs-linux/)
  , [Azure](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/expand-disks).

- If you use logging-machine(s) already in your cluster, it's necessary to scale up those machines before running
  upgrade to ensure you've got enough resources to run ELK stack in newer version. We recommend to use at least DS2_v2

>   Azure size (2 CPUs, 7 GB RAM) machine, or it's equivalent on AWS and on-prem installations. It's very related to

Reviewer: ```suggestion
  Azure size (2 CPUs, 7 GB RAM) machine, or its equivalent on AWS and on-prem installations. It's very related to
```

epicli upgrade -b /buildoutput/
```

This will backup and upgrade the Ansible inventory in the provided build folder `/buildoutput/` which will be used to
perform the upgrade of the components.

### Offline upgrade

#### Offline prerequisites

Your airgapped existing cluster should meet the following requirements:

1. The airgapped cluster machines/vm`s are connected by a network or virtual network of some sorts and can communicate

>    which each other:

Reviewer: ```suggestion
   with each other:
```

    - Has access to the build output from when the cluster was first created.
    - Is on the same network as your cluster machines
    - Has Epicli 0.4.2 or up running.

---
**NOTE**

> To run Epicli check the [Prerequisites](./PREREQUISITES.md)

Reviewer: ```suggestion
Before running `epicli`, check the [Prerequisites](./PREREQUISITES.md)
```

    epicli prepare --os OS
    ```

   Where OS should be `centos-7`, `redhat-7`, `ubuntu-18.04`. This will create a directory called `prepare_scripts` with
   the following files inside:

    - download-requirements.sh
    - requirements.txt

> 2. The scripts in the `prepare_scripts` will be used to download all requirements. To do that copy the `prepare_scripts`

Reviewer: ```suggestion
2. The scripts in the `prepare_scripts` will be used to download all requirements. To do that, copy the `prepare_scripts`
```

information visit RabbitMQ [site](https://www.rabbitmq.com/upgrade.html).

---

With the latest Epiphany version it's possible to upgrade RabbitMQ to v3.8.9. It requires Erlang system packages upgrade
that is done automatically to v23.1.4. Upgrade is performed in offline mode after stopping all RabbitMQ nodes.
[Rolling upgrade](https://www.rabbitmq.com/upgrade.html#rolling-upgrades) is not supported by Epiphany, and it is
advised not to use this approach when Erlang needs to be upgraded.

### Kubernetes upgrade

#### Prerequisites

Before K8s version upgrade make sure that deprecated API versions are not used:

1. [v1.17](https://v1-17.docs.kubernetes.io/docs/setup/release/notes/#deprecations-and-removals)
2. [v1.18](https://v1-18.docs.kubernetes.io/docs/setup/release/notes/#deprecation)

#### Upgrade

---
**NOTE**

If the K8s cluster that is going to be upgraded has the Istio control plane application deployed, issues can occur. The
default [profiles](https://istio.io/latest/docs/setup/additional-setup/config-profiles/) we currently support for
installing Istio only deploy a single replica for the control services with a `PodDisruptionBudgets` value of 0. This

> will result in the following error while its draining a these pods during an upgrade:

Reviewer: ```suggestion
will result in the following error while draining pods during an upgrade:
```

```shell
Cannot evict pod as it would violate the pods disruption budget.
```

As we currently don't support any kind of advanced configuration of the Istio control plane components outside the
default profiles, we need to scale up all components manually before the upgrade. This can be done with the following
command:

```shell
kubectl scale deploy -n istio-system --replicas=2 --all 
```

After the upgrade, the deployments can be scaled down to the original capacity:

```shell
kubectl scale deploy -n istio-system --replicas=1 --all 
```

**Note: The ```istio-system``` namespace value is the default value and should be set to whatever is being used in the
Istio application configuration.**

---

### PostgreSQL upgrade

---
**NOTE**

Before upgrade procedure, make sure you have a data backup.

---

#### Versions

With the latest Epiphany version it's possible to upgrade PostgreSQL to v13 with new extension versions:

- PgAudit v1.5.0
- repmgr v5.2.1

#### Upgrade

> Upgrade procedure is based on an [PostgreSQL documentation](https://www.postgresql.org/docs/13/pgupgrade.html) and

Reviewer: ```suggestion
Upgrade procedure is based on [PostgreSQL documentation](https://www.postgresql.org/docs/13/pgupgrade.html) and
```

requires downtime as there is a need to stop old service(s) and start new one(s).

There is a possibility to provide a custom configuration for upgrade with `epicli upgrade -f`, and there are a few
limitations related to specifying parameters for upgrade:

- If there were non-default values provided for installation (`epicli apply`), they have to be used again not to be
  overwritten by defaults.

- `wal_keep_segments` parameter for replication is replaced

>   by [wal_keep_size](https://www.postgresql.org/docs/13/runtime-config-replication.html#GUC-WAL-KEEP-SIZE) with a

Reviewer: ```suggestion
  by [wal_keep_size](https://www.postgresql.org/docs/13/runtime-config-replication.html#GUC-WAL-KEEP-SIZE) with the
```

  default value of 500 MB. Previous parameter is not supported.

- `archive_command` parameter for replication was set to `/bin/true` by default. It was planned to disable archiving,

>   but changes to `archive_mode` require a full PostgreSQL server restart, while archive_command changes can be applied

Reviewer: ```suggestion
  but changes to `archive_mode` require a full PostgreSQL server restart, while `archive_command` changes can be applied
```

  via a normal configuration reload. See [documentation](https://repmgr.org/docs/repmgr.html#CONFIGURATION-POSTGRESQL).

- There is no possibility to disable an extension after installation, so `specification.extensions.*.enabled: false`
  value will be ignored during upgrade if it was set to `true` during installation.

#### Post-upgrade processing

As p.13 of an [PostgreSQL documentation](https://www.postgresql.org/docs/13/pgupgrade.html) states there might be some
scripts generated after an upgrade. There is no clear description in which cases and what location they are created, so

> please check logs after an upgrade to see if additional steps required.

Reviewer: ```suggestion
please check logs after the upgrade to see if additional steps are required.
```
